[Concurrent Segment Search] Terms Agg Doc Count Error needs to be calculated across shards instead of slices

Subtask for #7357 to focus on doc count error issues

https://opensearch.org/docs/2.9/aggregations/bucket/terms/

```
REPRODUCE WITH: ./gradlew 'null' --tests "org.opensearch.search.aggregations.bucket.TermsDocCountErrorIT.testFixedDocs" -Dtests.seed=5B6F53240BBECC6D -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fi -Dtests.timezone=Jamaica -Druntime.java=11

java.lang.AssertionError: 
Expected: <46L>
     but: was <6L>
Expected :<46L>
Actual   :<6L>
<Click to see difference>


	at __randomizedtesting.SeedInfo.seed([5B6F53240BBECC6D:91411370660E073B]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:964)
	at org.junit.Assert.assertThat(Assert.java:930)
	at org.opensearch.search.aggregations.bucket.TermsDocCountErrorIT.testFixedDocs(TermsDocCountErrorIT.java:1098)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

```

--- a/server/src/main/java/org/opensearch/search/aggregations/bucket/terms/InternalTerms.java
+++ b/server/src/main/java/org/opensearch/search/aggregations/bucket/terms/InternalTerms.java
@@ -297,9 +297,11 @@ public abstract class InternalTerms<A extends InternalTerms<A, B>, B extends Int
         return thisReduceOrder != null ? thisReduceOrder : order;
     }
 
-    private long getDocCountError(InternalTerms<?, ?> terms) {
+    private long getDocCountError(InternalTerms<?, ?> terms, ReduceContext reduceContext) {
         int size = terms.getBuckets().size();
-        if (size == 0 || size < terms.getShardSize() || isKeyOrder(terms.order)) {
+        // doc_count_error is always computed at the coordinator based on the buckets returned by the shards. This should be 0 during the
+        // shard level reduce as no buckets are being pruned at this stage.
+        if (reduceContext.isSliceLevel() || size == 0 || size < terms.getShardSize() || isKeyOrder(terms.order)) {
             return 0;
         } else if (InternalOrder.isCountDesc(terms.order)) {
             if (terms.getDocCountError() > 0) {
@@ -412,7 +414,7 @@ public abstract class InternalTerms<A extends InternalTerms<A, B>, B extends Int
                 );
             }
             otherDocCount += terms.getSumOfOtherDocCounts();
-            final long thisAggDocCountError = getDocCountError(terms);
+            final long thisAggDocCountError = getDocCountError(terms, reduceContext);
             if (sumDocCountError != -1) {
                 if (thisAggDocCountError == -1) {
                     sumDocCountError = -1;

